### ISSUE TITLE
Logitech G-Keys showing as F-Keys (Logitech G910)

### ISSUE THREAD

**Issue by Friday13th87** [sezanzeb/key-mapper#187]
Hello there!
I tried to reassign my 9 G-Keys on my Logitech G910.
The Keys are assigned to F1-F9 in the hardware-profile it seems.
All i am getting are the F Keys when i hit the G Key in key-mapper, so i could reassign the F Keys but will loose them if i do so (so now gain of extra keys in the end)

Does anyone know how i could change that behavior or how i could make a work-around for that?

[on my logitech G604 mice i can program hardware profiles under the windows software g hub, and those keys are popping up then. couldnt find it for my keyboard
my goal was to have G1-G9 assigned to for example F13-21, so i can use them as extra keys]
thanks!

**Comment by sezanzeb** [comment]
I have trouble understanding the issue

> 9 G-Keys

> the G Key

So are there 9 G-Keys or is there only one?

> but will loose them if i do so (so now gain of extra keys in the end)

You loose keys, but at the same time gain extra keys?

**Comment by Friday13th87** [comment]
The Logitech G910 has 9 extra keys called "g-keys" - G1 to G9.
ootb in linux i am getting F1 by pressing G1, F2 by pressing G2 ... F9 by pressing G9.
In key-mapper i get the F(1-9) keys when trying to reassign the G(1-9) keys.
So if i reassign one of the G keys, which are originally assignt to the correspondenting F Keys, i will loose the function of the F keys itself, so i will loosing the original function of them, for example quick save and quick load (F5 and F6).
 What i thought Key-Mapper is actually doing, is assigning a specific function to the actual hardware key, not the output of that key.

**Comment by sezanzeb** [comment]
> What i thought Key-Mapper is actually doing, is assigning a specific function to the actual hardware key, not the output of that key.

no, key-mapper maps the events that arrive over usb. It doesn't modify/configure the hardware firmware

**Comment by Friday13th87** [comment]
thats not what i said.
there is a difference in the input between, for example, my G1 and my F1 key, even if both keys are doing the same. Both giving F1 as output, but the G1 key is another hardware button only that this is in the internal keyboard profile mapped to F1 - nevertheless this information is submitted as well.
Key-Mapper is not getting the actual usb hardware key, only the output signal its sending.
i found https://github.com/JSubelj/g910-gkey-macro-support, and this is actually directly binding the logitech g-keys, but key-mapper would have been more comfortable.

**Comment by sezanzeb** [comment]
Can you please generate some `sudo evtest` output for G1 and F1? I want to see if there is some difference that key-mapper might be able to detect

**Comment by Friday13th87** [comment]
bope its exactly the same in evtest, tried every G button:

`F2:
Event: time 1633377680.053773, -------------- SYN_REPORT ------------
Event: time 1633377703.501587, type 4 (EV_MSC), code 4 (MSC_SCAN), value 7003b
Event: time 1633377703.501587, type 1 (EV_KEY), code 60 (KEY_F2), value 1
Event: time 1633377703.501587, -------------- SYN_REPORT ------------
^[OQEvent: time 1633377703.607604, type 4 (EV_MSC), code 4 (MSC_SCAN), value 7003b
Event: time 1633377703.607604, type 1 (EV_KEY), code 60 (KEY_F2), value 0
Event: time 1633377703.607604, -------------- SYN_REPORT ------------

G2:
Event: time 1633377733.655994, type 4 (EV_MSC), code 4 (MSC_SCAN), value 7003b
Event: time 1633377733.655994, type 1 (EV_KEY), code 60 (KEY_F2), value 1
Event: time 1633377733.655994, -------------- SYN_REPORT ------------
^[OQEvent: time 1633377733.761943, type 4 (EV_MSC), code 4 (MSC_SCAN), value 7003b
Event: time 1633377733.761943, type 1 (EV_KEY), code 60 (KEY_F2), value 0
Event: time 1633377733.761943, -------------- SYN_REPORT ------------`

as it is working now with the gkey-macro application, i found a workaround how to get those extra keys working, even if the binding with your program would be much better.

**Comment by sezanzeb** [comment]
ok, unfortunate. Thanks for checking. I'm afraid it will be impossible for key-mapper to do anything there

**Comment by Friday13th87** [comment]
thanks for your quick response.
that g910-macro app can identify the G-keys separately.
didnt had the time to dig into it, to check how it is doing it.

**Comment by sezanzeb** [comment]
The only option might be to load a profile on the device itself that maps G1-G9 to other rarely used keys with some logitech driver, which is usually probably only possible in windows unless there is something developed by the community. After that key-mapper could map those keys without affecting the existing F* keys.

There is https://github.com/JSubelj/g910-gkey-macro-support/blob/e7515496ae8de858e870f3aa9fdc8e237f935e51/lib/g910_gkey_mapper.py#L96, which apparently is where the key is detected. I don't know if this is some reverse engineered proprietary protocol or some standard that key-mapper could use.
